### ISSUE TITLE
Command prefix support

### ISSUE THREAD

**Issue by bitprophet** [pyinvoke/invoke#293]
tl;dr This is Fabric 1's `sudo()`, `with prefix():`, and all the other related things which involve manipulating command strings given to `run()`.

For example, using a virtualenv to run Python-driven programs might look like this (yes, it's a bit of a devops antipattern, but it's a convenient example):

``` python
@task
def deploy(c):
    if not c.run("test -x myenv/bin/activate", hide=True):
        c.run("virtualenv myenv")
    with c.prefix("source myenv/bin/activate && "):
        # These all run as e.g. "source myenv/bin/activate && pip install ..."
        c.run("pip install -U -r requirements.txt")
        c.run("manage.py test")
        c.run("manage.py syncdb")
    # No prefixing occurs after this point
```

This could be implemented via `Context` state-keeping - yet another spot where we use them like Fabric 1's module global state. I thought about yielding a "new", modified context in the `with` block, but that stops making sense in Fabric 2 when each "context" is actually a `Connection`. And feels confusing and overwrought even in the base Invoke case, for little gain.

---

It feels like same implementation handling prefixes could also handle things that are sometimes-but-not-always prefix-driven, like `with cd(xxx):` which can be effected by `os.chdir` in Invoke but needs to be driven by `cd xxx &&` in Fabric's case & anything else remote-related. So it needs to be easy for other methods to update the list/whatever of prefixes.

---

Wrapping might also be nice, but actually may not be required up front (goal right now is to make `sudo` possible). Wrapping is needed for things like Fabric 1's shell-wrapping (turning `command` into `/bin/bash -l -c "command"`) but neither Invoke nor Fab 2 should actually need that in particular. However I could see users having cases where they do want to do something similar. If/when implemented it would want to use the same mechanisms as prefix control.

**Comment by chirhotec** [comment]
Are there still plans to implement this?

I'm actually trying to set up my tasks to run within an virtual environment (not sure why you say its an antipattern, could you explain?), and I'm trying to use virtualenvwrapper.

As per the virtualenvwrapper documentation, I've added `source <path/to/virtualenvwrapper.sh>` to my bash script, but the ctx.run() doesn't seem to be able to find the workon, mkvirtualenv, etc. commands.

To get around this, I currently have a helper function that prefixes the command:

```
def get_prefixed_command(command, prefix=None):
    return command if not prefix else "{} && {}".format(prefix, command)

def run(ctx, command, prefix=None, **kwargs):
    ctx.run(get_prefixed_command(command, prefix), **kwargs)
```

and then additional helper commands that prefix the `source <path/to/virtualenvwrapper.sh>` and `workon <venv_name>` commands:

```
def run_with_bash_script(ctx, command, prefix=None, **kwargs):
    return utils.run(
        ctx,
        utils.get_prefixed_command(command, prefix),
        prefix="source {}".format(ctx.config.get(BASH_SCRIPT_NAME_SETTING))
        )

def run_with_venv(ctx, command, prefix=None, **kwargs):
    return run_with_bash_script(
        ctx,
        utils.get_prefixed_command(command, prefix),
        prefix="workon {}".format(ctx.config.get(VIRTUAL_ENV_NAME_SETTING))
        )
```

Seems like a pretty ugly solution to me. Am I missing an easier solution?

**Comment by bitprophet** [comment]
First off, yes, I do still want to implement something along these lines eventually.

Re: virtualenvs specifically, you are missing a useful trick, and that's also why I noted the `source` option is an "antipattern"!

Command runners like Invoke need to run single-command shell sessions in order to present a useful programmatic API. Human-facing "source/activate a thing to mutate your shell environment, then leverage that mutation to actually execute the real command" don't work well with this; and they're just generally not as easy to script or automate in any tooling, being more brittle and easier to break.

The good news is that you don't actually need `source` or `workon` or etc; Python virtualenvs install all of their 'binary stubs' (pip, and other programs installed by Python packages, eg sphinx, invoke itself, etc) in a manner that references the virtualenv's specific Python interpreter. You can invoke them directly via their path, and they "just work".

So, given the prior execution of `mkvirtualenv foobar` (or, more directly, `virtualenv /explicit/path/to/an/environment/directory` - though as long as you know where `virtualenvwrapper` is putting its envs, it doesn't really matter), a human would do this:

```
workon foobar
pip install somepythonprogram
somepythonprogram --options
```

What a script wants to do is simply:

```
/path/to/env/bin/pip install somepythonprogram
```

followed by:

```
/path/to/env/bin/somepythonprogram --options
```

I broke them up to highlight the fact that the two commands do not require being in the same shell environment/session. All state is on-disk, in the virtualenv. Hope that all makes sense!

**Comment by bitprophet** [comment]
Unsure how we missed this ticket when we did #446 back around PyCon. `Context.prefix` is in!
